src/framework/results/experiment_data.hpp

```diff
@@ -177,205 +246,357 @@ void ExperimentData::add_memory_count(const std::string &memory) {
+
+void ExperimentData::check_reserved_key(const std::string &key) {
+  if (reserved_keys_.find(key) != reserved_keys_.end()) {
+    throw std::invalid_argument(
+        "Cannot add additional data with reserved key name \"" + key + "\".");
   }
 }
 
+void ExperimentData::erase_additional_data(const std::string &key) {
+  additional_json_data_.erase(key);
+  additional_cvector_data_.erase(key);
+  additional_cmatrix_data_.erase(key);
+}
 
 template <typename T>
-void ExperimentData::add_additional_data(const std::string &key, const T &data) {
+void ExperimentData::add_additional_data(const std::string &key, T &&data) {
+  check_reserved_key(key);
+  if (return_additional_data_) {
+    json_t jdata = data;
+    add_additional_data(key, std::move(jdata));
+  }
+}
+
+template <>
+void ExperimentData::add_additional_data(const std::string &key,
+                                         json_t &&data) {
+  check_reserved_key(key);
   if (return_additional_data_) {
-    json_t js = data; // use implicit to_json conversion function for T
-    if (JSON::check_key(key, additional_data_))
-      additional_data_[key].update(js.begin(), js.end());
-    else
-      additional_data_[key] = js;
+    erase_additional_data(key);
+    additional_json_data_[key] = data;
   }
 }
 
+template <>
+void ExperimentData::add_additional_data(const std::string &key,
+                                         cvector_t &&data) {
+  check_reserved_key(key);
+  if (return_additional_data_) {
+    erase_additional_data(key);
+    additional_cvector_data_[key] = data;
+  }
+}
 
-void ExperimentData::clear_additional_data(const std::string &key) {
-  additional_data_.erase(key);
+template <>
+void ExperimentData::add_additional_data(const std::string &key,
+                                         cmatrix_t &&data) {
+  check_reserved_key(key);
+  if (return_additional_data_) {
+    erase_additional_data(key);
+    additional_cmatrix_data_[key] = data;
+  }
 }
 
+//------------------------------------------------------------------
+// Metadata
+//------------------------------------------------------------------
 
 template <typename T>
-void ExperimentData::add_metadata(const std::string &key, const T &data) {
-  json_t js = data; // use implicit to_json conversion function for T
-  if (JSON::check_key(key, additional_data_))
-    metadata_[key].update(js.begin(), js.end());
-  else
-    metadata_[key] = js;
+void ExperimentData::add_metadata(const std::string &key, T &&data) {
+  // Use implicit to_json conversion function for T
+  json_t jdata = data;
+  add_metadata(key, std::move(jdata));
 }
 
-
-void ExperimentData::clear_metadata(const std::string &key) {
-  metadata_.erase(key);
+template <>
+void ExperimentData::add_metadata(const std::string &key, json_t &&data) {
+  auto elt = metadata_.find("key");
+  if (elt == metadata_.end()) {
+    // If key doesn't already exist add new data
+    metadata_[key] = data;
```

Comment:
```suggestion
    metadata_[key] = std::move(data);
```
... because `metadata_[key]` is a `json_t` which implements move semantics